### PR TITLE
Majordomo Broker V0.1 PoC

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
Majordomo Broker V0.1 proof-of-concept to be developed into an independent microservice middle-ware

This is the reference implementation to compare future iterations and upgrades against.

### CODE REVIEW

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 22
           

Complexity increasing per file
==============================
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java  15
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java  15
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java  11
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/SimpleEchoServiceWorker.java  1
- microservice/src/test/java/de/gsi/microservice/utils/SystemProperties.java  4
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoProtocol.java  22
- microservice/src/test/java/de/gsi/microservice/utils/WorkerThreadFactory.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV1.java  9
- microservice/src/test/java/de/gsi/microservice/rbac/BasicRbacRole.java  1
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV1.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV2.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java  4
- microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyTests.java  19
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  12
- microservice/src/test/java/de/gsi/microservice/rbac/RbacToken.java  5
         

Clones added
============
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742&bid=20511185)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+    // ---------------------------------------------------------------------
+    private static final int BLOCKING_QUEUE_DEFAULT_SIZE = 10;
+    protected final String uniqueID;
+    protected final ZContext ctx;
+    private final String brokerAddress;
+    private final String serviceName;
+    private final byte[] serviceBytes;
+
+    private final AtomicBoolean run = new AtomicBoolean(true);
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java:53]: Codacy found an issue: [Avoid unused private fields such as 'run'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+ *
+ */
+public class MajordomoWorker extends Thread {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MajordomoWorker.class);
+    private static final int HEARTBEAT_LIVENESS = SystemProperties.getValueIgnoreCase("OpenCMW.heartBeatLiveness", 3); // [counts] 3-5 is reasonable
+    private static final int HEARTBEAT_INTERVAL = SystemProperties.getValueIgnoreCase("OpenCMW.heartBeat", 2500); // [ms]
+    private static final AtomicInteger WORKER_COUNTER = new AtomicInteger();
+
+    // ---------------------------------------------------------------------
+    private static final int BLOCKING_QUEUE_DEFAULT_SIZE = 10;
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java:46]: Codacy found an issue: [Avoid unused private fields such as 'BLOCKING_QUEUE_DEFAULT_SIZE'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+        sendHeartBeat();
+
+        get("testdevice", "unknownProp", "FAIR.SELECTOR.ALL");
+
+        // return reply to data
+        while (true) {
+            final byte[] data = controlChannel.recv();
+            if (data != null && data.length != 0) {
+                if (Arrays.equals(data, new byte[] { MT_HEADER, MT_BODY_EXCEPTION })) { // header, exception
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java:211]: Codacy found an issue: [These nested if statements could be combined](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+    public void removeService(final String serviceName) {
+        services.remove(serviceName);
+    }
+
+    /**
+     * main broker work happens here
+     */
+    public void run() {
+        final ZMQ.Poller items = ctx.createPoller(routerSockets.size());
+        for (ZMQ.Socket routerSocket : routerSockets) {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java:153]: Codacy found an issue: [Unnecessary use of fully qualified name 'ZMQ.Socket' due to existing static import 'org.zeromq.ZMQ.Socket'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    @Override
+    public synchronized void start() {
+        reconnectToBroker();
+        super.start();
+    }
+
+    /**
+     * Send reply, if any, to broker and wait for next request.
+     */
+    protected void handleReceive() {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java:203]: Codacy found an issue: [The method 'handleReceive()' has an NPath complexity of 554, current threshold is 200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+                if (client.protocol == MdpSubProtocol.C_CLIENT) { // OpenCMW
+                    MajordomoProtocol.sendClientMessage(client.socket, MdpClientCommand.C_UNKNOWN, msg.clientSourceID, msg.serviceNameBytes, msg.payload);
+                } else {
+                    // TODO: add other branches for:
+                    // * CmwLight
+                    // * REST/JSON
+                    // * REST/HTML
+                    throw new IllegalStateException("Unexpected value: " + client.protocol);
+                }
+                if (receiveSocket != internalServiceSocket) {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java:333]: Codacy found an issue: [Use equals() to compare object references.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/utils/SystemProperties.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
+package de.gsi.microservice.utils;
+
+import static java.util.Map.Entry;
+
+import java.util.Properties;
+import java.util.Set;
+
+public class SystemProperties {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/utils/SystemProperties.java:8]: Codacy found an issue: [The utility class name 'SystemProperties' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    // ---------------------------------------------------------------------
+    private static final int BLOCKING_QUEUE_DEFAULT_SIZE = 10;
+    protected final String uniqueID;
+    protected final ZContext ctx;
+    private final String brokerAddress;
+    private final String serviceName;
+    private final byte[] serviceBytes;
+
+    private final AtomicBoolean run = new AtomicBoolean(true);
+    private final SortedSet<RbacRole<?>> rbacRoles;
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java:54]: Codacy found an issue: [Avoid unused private fields such as 'rbacRoles'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+        if (oldWorker != null) {
+            LOGGER.atWarn().addArgument(worker.getServiceName()).log("overwriting existing internal service definition for '{}'");
+        }
+    }
+
+    /**
+     * Bind broker to endpoint, can call this multiple times. We use a single
+     * socket for both clients and workers.
+     */
+    public ZMQ.Socket bind(String endpoint) {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java:114]: Codacy found an issue: [Unnecessary use of fully qualified name 'ZMQ.Socket' due to existing static import 'org.zeromq.ZMQ.Socket'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoProtocol.java [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+import org.zeromq.util.ZData;
+
+import zmq.SocketBase;
+
+/**
+ * Majordomo Protocol (MDP) definitions and implementations according to https://rfc.zeromq.org/spec/7/
+ *
+ * @author rstein
+ */
+public class MajordomoProtocol {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoProtocol.java:21]: Codacy found an issue: [The utility class name 'MajordomoProtocol' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+                break;
+            }
+        }
+    }
+
+    private static void handleStreamHttpSocket(Socket httpSocket, ZFrame handlerExt) {
+        // Get [id, ] message on client connection.
+        ZFrame handler = handlerExt;
+        if (handler == null) {
+            if ((handler = getConnectionID(httpSocket)) == null) {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java:174]: Codacy found an issue: [These nested if statements could be combined](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java [hunk 12]
```diff
@@ -0,0 +1,648 @@
...
+        final ZMQ.Socket routerSocket = ctx.createSocket(SocketType.ROUTER);
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java:115]: Codacy found an issue: [Unnecessary use of fully qualified name 'ZMQ.Socket' due to existing static import 'org.zeromq.ZMQ.Socket'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java [hunk 13]
```diff
@@ -0,0 +1,648 @@
...
+    private final Deque<Worker> waiting = new ArrayDeque<>(); // idle workers
+    private long heartbeatAt = System.currentTimeMillis() + HEARTBEAT_INTERVAL; // When to send HEARTBEAT
+
+    /**
+     * Initialize broker state.
+     * @param ioThreads number of threads dedicated to network IO (recommendation 1 thread per 1 GBit/s)
+     * @param nInternalThreads number of threads per internal service
+     * @param rbacRoles RBAC-based roles (used for IO prioritisation and service access control
+     */
+    public MajordomoBroker(final int ioThreads, final int nInternalThreads, final RbacRole<?>... rbacRoles) {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java:85]: Codacy found an issue: [Avoid unused constructor parameters such as 'nInternalThreads'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyTests.java [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+        }
+
+        final long stop = System.currentTimeMillis();
+        System.out.printf("%-40s:  %10d calls/second\n", topic, (1000 * nExec) / (stop - start));
+    }
+
+    static class Broker implements Runnable {
+        private static final int TIMEOUT = 1000;
+        @Override
+        public void run() {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyTests.java:100]: Codacy found an issue: [The method 'run()' has an NPath complexity of 2606, current threshold is 200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java [hunk 15]
```diff
@@ -0,0 +1,648 @@
...
+    }
+
+    public ZMQ.Socket getInternalRouterSocket() {
+        return internalRouterSocket;
+    }
+
+    /**
+     * @return unmodifiable list of registered external sockets
+     */
+    public List<ZMQ.Socket> getRouterSockets() {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java:136]: Codacy found an issue: [Unnecessary use of fully qualified name 'ZMQ.Socket' due to existing static import 'org.zeromq.ZMQ.Socket'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+        } else if (Arrays.equals(data.getLast().getData(), new byte[] { MT_HEADER, MT_BODY_EXCEPTION })) {
+            System.out.println("received exception: " + data);
+        } else if (!Arrays.equals(data.getLast().getData(), new byte[] { MT_HEADER, MT_BODY, MT_BODY_DATA_CONTEXT })) {
+            System.out.println("expected reply message, but got: " + Arrays.toString(data.getLast().getData()));
+            return;
+        } else {
+            System.out.println("received reply: " + data);
+        }
+        for (final byte desc : descriptor) {
+            switch (desc) {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java:144]: Codacy found an issue: [A switch statement does not contain a break](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java [hunk 17]
```diff
@@ -0,0 +1,648 @@
...
+                dispatch(service);
+            }
+        });
+    }
+
+    /**
+     * Process message sent to us by a worker.
+     */
+    protected void processWorker(final Socket receiveSocket, final MdpWorkerMessage msg) {
+        boolean workerReady = receiveSocket == internalServiceSocket || workers.containsKey(msg.senderIdHex);
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java:304]: Codacy found an issue: [Use equals() to compare object references.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java (hunk 8 above)
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java:54]: Codacy found an issue: [Perhaps 'rbacRoles' could be replaced by a local variable.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyTests.java (hunk 14 above)
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyTests.java:100]: Codacy found an issue: [Avoid really long methods.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    /**
+     * Send request to broker and get reply by hook or crook Takes ownership of request message and destroys it when sent.
+     *
+     * @param service UTF-8 encoded service name byte array
+     * @param msgs message(s) to be sent to MajordomoProtocol broker (if more than one, than the last is assumed to be a RBAC-token
+     */
+    public boolean send(final byte[] service, final byte[]... msgs) {
+        return MajordomoProtocol.sendClientMessage(clientSocket, MdpClientCommand.C_UNKNOWN, null, service, msgs);
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java:119]: Codacy found an issue: [Unnecessary use of fully qualified name 'MajordomoProtocol.sendClientMessage' due to existing static import 'de.gsi.microservice.concepts.majordomo.MajordomoProtocol.*'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java [hunk 19]
```diff
@@ -0,0 +1,648 @@
...
+            LOGGER.atDebug().addArgument(endpoint).log("Majordomo broker/0.1 is active at '{}'");
+        }
+        return routerSocket;
+    }
+
+    public ZContext getContext() {
+        return ctx;
+    }
+
+    public ZMQ.Socket getInternalRouterSocket() {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java:129]: Codacy found an issue: [Unnecessary use of fully qualified name 'ZMQ.Socket' due to existing static import 'org.zeromq.ZMQ.Socket'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+            handle.send(socket, ZFrame.MORE | ZFrame.REUSE);
+            socket.send((byte[]) null, 0);
+            return null;
+        }
+
+        // receive request
+        return ZFrame.recvFrame(socket);
+    }
+
+    private static void handleRouterSocket(final Socket router, final Socket stream) {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java:129]: Codacy found an issue: [Avoid unused method parameters such as 'stream'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/SimpleEchoServiceWorker.java  1
- microservice/src/test/java/de/gsi/microservice/utils/WorkerThreadFactory.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoProtocol.java  22
- microservice/src/test/java/de/gsi/microservice/rbac/BasicRbacRole.java  1
- microservice/src/test/java/de/gsi/microservice/utils/SystemProperties.java  4
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  14
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java  15
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV1.java  9
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  5
- microservice/src/test/java/de/gsi/microservice/rbac/RbacToken.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java  4
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV2.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyEvaluation.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java  11
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV1.java  6
         

Clones added
============
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742&bid=20511185)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java [hunk 21]
```diff
@@ -0,0 +1,328 @@
...
+        } catch (UnknownHostException e) {
+            hostname = "localhost";
+        }
+        final long processId = ProcessHandle.current().pid();
+        final long connectionId = 1L;
+        final long channelId = 1L;
+        return hostname + '/' + processId + '/' + connectionId + '/' + channelId;
+    }
+
+    private void receiveData() {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java:120]: Codacy found an issue: [The method 'receiveData()' has an NPath complexity of 448, current threshold is 200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java:129]: Codacy found an issue: [Avoid unused method parameters such as 'stream'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV1.java  6
- microservice/src/test/java/de/gsi/microservice/rbac/RbacToken.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/SimpleEchoServiceWorker.java  1
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoProtocol.java  22
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV1.java  9
- microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java  11
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV2.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyEvaluation.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java  15
- microservice/src/test/java/de/gsi/microservice/utils/WorkerThreadFactory.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java  4
- microservice/src/test/java/de/gsi/microservice/utils/SystemProperties.java  4
- microservice/src/test/java/de/gsi/microservice/rbac/BasicRbacRole.java  1
         

Clones added
============
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742&bid=20511185)

**Review by wirew0rm — DISMISSED** [review]
LGTM :+1:

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/SimpleEchoServiceWorker.java [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+
+import de.gsi.microservice.rbac.BasicRbacRole;
+
+/**
+* Majordomo Protocol worker example. Uses the mdwrk API to hide all MajordomoProtocol aspects
+*
+*/
+public class SimpleEchoServiceWorker {
+    public static void main(String[] args) {
+        boolean verbose = (args.length > 0 && "-v".equals(args[0]));
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/SimpleEchoServiceWorker.java:11]: Codacy found an issue: [Avoid unused local variables such as 'verbose'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBrokerTests.java  1
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV2.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java  3
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoProtocol.java  22
- microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java  11
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV1.java  9
- microservice/src/test/java/de/gsi/microservice/utils/SystemProperties.java  4
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV1.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  5
- microservice/src/test/java/de/gsi/microservice/rbac/RbacToken.java  5
- microservice/src/test/java/de/gsi/microservice/rbac/BasicRbacRole.java  1
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/SimpleEchoServiceWorker.java  1
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java  19
- microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyEvaluation.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  6
- microservice/src/test/java/de/gsi/microservice/utils/WorkerThreadFactory.java  6
         

Clones added
============
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742&bid=20511185)

File: microservice/src/test/java/de/gsi/microservice/utils/WorkerThreadFactory.java [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+        this.poolName = poolName;
+        this.nThreads = nThreads <= 0 ? MAX_THREADS : nThreads;
+        this.pool = Executors.newFixedThreadPool(this.nThreads, SELF == null ? this : SELF);
+        if (this.pool instanceof ThreadPoolExecutor) {
+            ((ThreadPoolExecutor) pool).setRejectedExecutionHandler((runnable, executor) -> {
+                try {
+                    // work queue is full -> make the thread calling pool.execute() to wait
+                    executor.getQueue().put(runnable);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/utils/WorkerThreadFactory.java:41]: Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBrokerTests.java [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+            assertNotNull(replyWithoutRbac, "reply message w/o RBAC token not being null");
+            assertNotNull(replyWithoutRbac.peekLast(), "user-data not being null");
+            assertEquals("501", new String(replyWithoutRbac.pollLast().getData()), "unknown service");
+        }
+
+        broker.stopBroker();
+    }
+
+    @Test
+    public void basicASynchronousRequestReplyTest() throws IOException {
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBrokerTests.java:227]: Codacy found an issue: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

File: microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+
+/**
+ * Majordomo Protocol Client API, asynchronous Java version. Implements the
+ * MajordomoProtocol/Worker spec at http://rfc.zeromq.org/spec:7.
+ */
+public class MajordomoClientV2 {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MajordomoClientV2.class);
+    private final String broker;
+    private final ZContext ctx;
+    private final Formatter log = new Formatter(System.out);
```
Comment (RalphSteinhagen) [inline microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java:24]: Codacy found an issue: [Avoid unused private fields such as 'log'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV2.java  3
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/SimpleEchoServiceWorker.java  1
- microservice/src/test/java/de/gsi/microservice/rbac/BasicRbacRole.java  1
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoWorker.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoClientV1.java  9
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoProtocol.java  22
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV2.java  7
- microservice/src/test/java/de/gsi/microservice/concepts/RestBehindRouterEvaluation.java  11
- microservice/src/test/java/de/gsi/microservice/utils/SystemProperties.java  4
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/ClientSampleV1.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBroker.java  19
- microservice/src/test/java/de/gsi/microservice/utils/WorkerThreadFactory.java  6
- microservice/src/test/java/de/gsi/microservice/rbac/RbacToken.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  5
- microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyEvaluation.java  6
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/MajordomoBrokerTests.java  1
         

Clones added
============
- microservice/src/test/java/de/gsi/microservice/concepts/majordomo/legacy/CmwLightClient.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/BlockingQueueTests.java  2
- microservice/src/test/java/de/gsi/microservice/concepts/PushPullTests.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742&bid=20511185)

**Review by wirew0rm — APPROVED** [review]
:+1: looks good

**Pull request merged**